# Clase de 13 de Octubre 2021

Manipulación de datos reales con Pandas

En esta parte de la clase utilizaremos folium, una descripción completa esta disponible [aquí](https://python-visualization.github.io/folium/)

In [1]:
import pandas as pd
import folium as mapas

#### Vamos a crear un mapa que tiene como centro la ciudad de Santiago de Chile con las coordenadas obtenidas en el [siguiente link](https://es.db-city.com/Chile--Santiago--de-Santiago--Santiago) utilizando folium

In [2]:
mapa = mapas.Map(location=[-33.4513,-70.6653],zoom_start=12,control_scale=True)

In [7]:
mapa

### Agregar un marcador para cada comuna  
  
Utilizaremos los datos de latitud y longitud de el siguiente repositorio de (github)[https://github.com/2x3-la/geo-chile]

In [5]:
df= pd.read_json("https://raw.githubusercontent.com/2x3-la/geo-chile/master/chile.json")

In [6]:
df

name        lng        lat
0         Santiago -70.666667 -33.450000
1      Cerro Navia -70.716667 -33.416667
2        El Bosque -70.700000 -33.566667
3       Huechuraba -70.666667 -33.350000
4      La Cisterna -70.683333 -33.550000
..             ...        ...        ...
340        Natales -72.516667 -51.733333
341      Primavera -69.250000 -52.716667
342      Antártica -71.500000 -75.000000
343   San Gregorio -69.683333 -52.316667
344  Laguna Blanca -71.916667 -52.250000

[345 rows x 3 columns]

In [16]:
df.shape

(345, 3)

In [15]:
df.describe()

lng         lat
count  345.000000  345.000000
mean   -71.674863  -35.591493
std      3.121541    7.529802
min   -109.375000  -75.000000
25%    -72.583333  -38.400000
50%    -71.583333  -34.983333
75%    -70.816667  -33.400000
max    -33.408148   34.287229

In [12]:
df.lat.mean()

-35.591493246666666

In [13]:
df.lng.mean()

-71.67486295391306

In [14]:
mapa = mapas.Map(location=[df.lat.mean(),df.lng.mean()],zoom_start=4,control_scale=True)
mapa

##### Vamos a poner un marcador para cada una de las 345 comunas

In [20]:
for indice, valores in df.iterrows():
    print(valores['name'])


Santiago
Cerro Navia
El Bosque
Huechuraba
La Cisterna
La Granja
La Reina
Lo Barnechea
Lo Prado
Maipú
Pedro Aguirre Cerda
Providencia
Quilicura
Recoleta
San Joaquín
San Ramón
Puente Alto
Padre Hurtado
El Monte
San Pedro
Curacaví
Melipilla
Calera de Tango
San Bernardo
Lampa
San José de Maipo
Peñaflor
Isla de Maipo
Talagante
María Pinto
Paine
Buin
Tiltil
Colina
Pirque
Vitacura
San Miguel
Renca
Quinta Normal
Pudahuel
Peñalolén
Ñuñoa
Macul
Lo Espejo
Las Condes
La Pintana
La Florida
Independencia
Estación Central
Conchalí
Cerrillos
Arica
Camarones
Putre
General Lagos
Iquique
Alto Hospicio
Pozo Almonte
Camiña
Colchane
Huara
Pica
Antofagasta
Mejillones
Sierra Gorda
Taltal
Calama
Ollague
San Pedro de Atacama
María Elena
Tocopilla
Copiapó
Caldera
Tierra Amarilla
Chañaral
Diego de Almagro
Vallenar
Alto del Carmen
Freirina
Huasco
Río Hurtado
Monte Patria
Ovalle
Los Vilos
Illapel
Paiguano
Andacollo
La Serena
Punitaqui
Combarbalá
Salamanca
Canela
Vicuña
La Higuera
Coquimbo
Valparaíso
Concón
Puchunca

'Laguna Blanca'

In [18]:
df['name']

0           Santiago
1        Cerro Navia
2          El Bosque
3         Huechuraba
4        La Cisterna
           ...      
340          Natales
341        Primavera
342        Antártica
343     San Gregorio
344    Laguna Blanca
Name: name, Length: 345, dtype: object

In [25]:
for indice, valores in df.iterrows():
    mapas.Marker([valores['lat'],valores['lng']],valores['name']).add_to(mapa)

In [24]:
for indice, valores in df.iterrows():
    mapas.Marker(location=[valores['lat'],valores['lng']],popup=valores['name']).add_to(mapa)

### Si tuvieran problemas porque son muchas comunas, 
### podemos guardar un nuevo dataframe con solo los valores de head=10

In [29]:
df2=df.head(10)

In [30]:
mapa = mapas.Map(location=[df2.lat.mean(),df2.lng.mean()],zoom_start=4,control_scale=True)
mapa

In [31]:
for indice, valores in df2.iterrows():
    mapas.Marker(location=[valores['lat'],valores['lng']],popup=valores['name']).add_to(mapa)

In [32]:
mapa

In [33]:
casos_confirmados = pd.read_csv("https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto2/2020-07-13-CasosConfirmados.csv")

In [36]:
casos_confirmados

Region  Codigo region                          Comuna  \
0    Arica y Parinacota             15                           Arica   
1    Arica y Parinacota             15                       Camarones   
2    Arica y Parinacota             15                   General Lagos   
3    Arica y Parinacota             15                           Putre   
4    Arica y Parinacota             15  Desconocido Arica y Parinacota   
..                  ...            ...                             ...   
357          Magallanes             12                       Rio Verde   
358          Magallanes             12                    San Gregorio   
359          Magallanes             12                        Timaukel   
360          Magallanes             12                Torres del Paine   
361          Magallanes             12          Desconocido Magallanes   

     Codigo comuna  Poblacion  Casos Confirmados  
0          15101.0   247552.0             2721.0  
1          15102.0     1233.0                8.0  
2          15202.0      810.0                0.0  
3          15201.0     2515.0               28.0  
4              NaN        NaN                3.0  
..             ...        ...                ...  
357        12103.0      211.0                0.0  
358        12104.0      681.0                2.0  
359        12303.0      282.0                0.0  
360        12402.0     1021.0                0.0  
361            NaN        NaN               70.0  

[362 rows x 6 columns]

In [40]:
casos_confirmados.shape

(362, 6)

In [41]:
casos_confirmados.describe()

Codigo region  Codigo comuna      Poblacion  Casos Confirmados
count     362.000000     346.000000     346.000000         362.000000
mean        8.784530    9034.997110   56237.890173         985.143646
std         3.884479    3818.147431   88945.967089        2493.129402
min         1.000000    1101.000000     137.000000           0.000000
25%         6.000000    6109.250000    9649.250000          33.250000
50%         8.000000    8313.500000   19770.000000         110.500000
75%        13.000000   13102.750000   55441.250000         427.750000
max        16.000000   16305.000000  645909.000000       20923.000000

In [39]:
casos_confirmados[columnas]

Comuna  Casos Confirmados
0                             Arica             2721.0
1                         Camarones                8.0
2                     General Lagos                0.0
3                             Putre               28.0
4    Desconocido Arica y Parinacota                3.0
..                              ...                ...
357                       Rio Verde                0.0
358                    San Gregorio                2.0
359                        Timaukel                0.0
360                Torres del Paine                0.0
361          Desconocido Magallanes               70.0

[362 rows x 2 columns]

In [ ]:
df.merge?

In [ ]:
columnas=["Comuna","Casos Confirmados"]

In [45]:
casos_confirmados=casos_confirmados[columnas]

In [46]:
df.merge(casos_confirmados, left_on=['name'], right_on=['Comuna'])

name        lng        lat            Comuna  \
0            Santiago -70.666667 -33.450000          Santiago   
1         Cerro Navia -70.716667 -33.416667       Cerro Navia   
2           El Bosque -70.700000 -33.566667         El Bosque   
3          Huechuraba -70.666667 -33.350000        Huechuraba   
4         La Cisterna -70.683333 -33.550000       La Cisterna   
..                ...        ...        ...               ...   
260  Torres del Paine -72.350000 -51.266667  Torres del Paine   
261           Natales -72.516667 -51.733333           Natales   
262         Primavera -69.250000 -52.716667         Primavera   
263      San Gregorio -69.683333 -52.316667      San Gregorio   
264     Laguna Blanca -71.916667 -52.250000     Laguna Blanca   

     Casos Confirmados  
0              13852.0  
1               5861.0  
2               5840.0  
3               3474.0  
4               2932.0  
..                 ...  
260                0.0  
261               51.0  
262                1.0  
263                2.0  
264                0.0  

[265 rows x 5 columns]